{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Usher Syndrome and the Evolution of Metazoan Sensory Structures\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "Usher Syndrome (USH), a genetic sensory disorder, is the most common cause of combined blindness-deafness in humans.  The genes associated with Usher syndrome play key structural and functional roles in ciliated sensory cells of the vertebrate retina and inner ear.  Usher genes form interciliary links and their anchoring complexes in photoreceptors and the mechanosensory hair cell (Kremer et al 2006).  When a mutation occurs in one of these genes,  mechanotransduction is abolished and the retina degenerates, resulting in blindness, deafness and impaired vestibular function.  \n",
    "\n",
    "Given the key role these genes play in vertebrate sensory structures, it is concievable that these genes may serve similar sensory functions in other Metazoan groups.   Previously thought to be confined to vertebrates, USH homologs were identifed within an Echinoderm genome (Sodergren 2006).  Recently, USH homologs have been shown to be upregulated in the choanocytes of the sponge *Ephydatia*, hinting that these genes may play a conserved role in the evolution of ciliated sensory structures of the Metazoa, and begging the question of how early these genes arose (Pena et al 2016).  By investigating the evolutionary history of the genes involved in Usher syndrome, this project can better determine how the suite of genes involved with Usher syndrome were assembled within the Metazoa and its close relatives, and what role these genes might have played in the sensory evolution of early animals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Steps to the process"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BLAST  Human USH sequences against NCBI data base\n",
    "\n",
    "```\n",
    "def search_taxa_all_gene_delay(list_of_taxa):\n",
    "    # blasts sequences in a file against a user submitted list of taxa\n",
    "    # loop through the list and run blast for each one\n",
    "    # will save each result to a separate xml file\n",
    "    \n",
    "    \n",
    "    from Bio.Blast import NCBIWWW\n",
    "        # imports the NCBIWWW module from Biopython to allow remote querry \n",
    "        \n",
    "    import time\n",
    "        # let's us delay imputs to not spam the NCBI servers and get kicked off\n",
    "    \n",
    "    with open(\"USH_Search_seq.fasta\", \"r\") as fasta_file:\n",
    "        sequences = fasta_file.read()\n",
    "        fasta_file.close()\n",
    "        #reads in sequences we will be searching\n",
    "        \n",
    "    for i in list_of_taxa:\n",
    "        result_handle = NCBIWWW.qblast(\"blastp\", # specifies the program for a protein-protein search\n",
    "                                       \"refseq_protein\",  #  database of protein sequences\n",
    "                                       sequences, # our list of sequences we read in\n",
    "                                       alignments = 100, # asks for 100 best hits\n",
    "                                       descriptions = 100, \n",
    "                                       expect = 0.00001, # specifies the E-value cut off (i.e. how likely a random match for our query would be\n",
    "                                       entrez_query = str(i)) # specifies the taxa as we loop through it\n",
    "                                       \n",
    "        file_name=str(\"USH_Search_\"+str(i)+\".xml\") #this creates a name for the file\n",
    "        \n",
    "        save_file=open(file_name, \"w\")  #we are opening a file that does not yet exist to write to it\n",
    "        \n",
    "        save_file.write(result_handle.read())  #writing the result of our blast search to local file\n",
    "        \n",
    "        save_file.close() #closing it to allow the file to actually write it\n",
    "        result_handle.close() #close the results handle\n",
    "        \n",
    "        print(\"created \"+ file_name) #this is just a nice way to track the progress of the program\n",
    "        \n",
    "        time.sleep(60)  #this gives 1 minute between writing the output and sending another request to the ncbi server\n",
    "            # NCBI is a shared resource, so you can't monopolize computer time\n",
    "        \n",
    "```\n",
    "\n",
    "\n",
    "Here is the list of taxa:\n",
    "```\n",
    "full_taxa_file_name=open(\"/home/eeb177-student/Desktop/eeb-177/project/sandbox/Testrun_multi_genes_same_org/full_list_taxa_NCBI.txt\", \"r\")\n",
    "```\n",
    "please note that this only includes all the taxa from NCBI's database, there are still a number I want to include in the analysis from separate databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse BLAST Output to CSV\n",
    "```\n",
    "def parse_and_summarize(blast_output_xml):\n",
    "    #  goes through the output of a BLAST xml file and finds the relevant stats to summarize the search\n",
    "    from Bio.Blast import NCBIXML\n",
    "    from Bio.SeqRecord import SeqRecord\n",
    "    #import the required libraries\n",
    "    \n",
    "    for file_name in blast_output_xml:\n",
    "        result_handle = open(str(file_name), \"r\") #sets the result handle\n",
    "        \n",
    "        blast_records = NCBIXML.parse(result_handle)\n",
    "        #need to use parse if it has multiple records in it\n",
    "        \n",
    "        for blast_record in blast_records:\n",
    "            org_desig=file_name.split(\"_\")[2].split(\"[\")[0].replace(\" \", \"_\")\n",
    "            #properly formats the taxa id so we can use it to name things\n",
    "            \n",
    "            homo_sapiens = \"[Homo sapiens]\"\n",
    "            blast_query=blast_record.query\n",
    "            if homo_sapiens in blast_query:\n",
    "                gene_name=blast_record.query.split(\"|\")[4].split(\"[\")[0].replace(\" \", \"_\").replace(\"_protein\", \"\").replace(\"_isoform_b3\", \"\")\n",
    "                formated_gene_name = gene_name[1:-1]\n",
    "            else:\n",
    "                formated_gene_name=blast_query.split(\"|\")[4].split(\"_\")[0]\n",
    "            #this conditional properly formats the gene name so we can use it to name things\n",
    "                #Basically there are 2 formats of sequence names I used to do the search\n",
    "                #this statement switches the naming convention depending on which is used\n",
    "                #it's not very general, so in the future I will make sure to properly name my search sequences\n",
    "                #that should make this statement unnecessary \n",
    "```\n",
    "This gives us the BLAST results for one gene in one species as a CSV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Grab Gene IDs from output with bash for each gene\n",
    "\n",
    "```\n",
    "cat gene_WHRN_tiny_subset_summary.csv | cut -d ',' -f 1 > WHRN_tiny_gi.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download full sequence for gene from NCBI\n",
    "```\n",
    "def download_seqs_from_list_and_autoname(in_filename):    \n",
    "    with open(in_filename, \"r\") as query_file:\n",
    "        query_ids = query_file.read().splitlines()\n",
    "    #open the file, grab the ids\n",
    "\n",
    "    from Bio import Entrez\n",
    "    Entrez.email = \"hspeck@ucla.edu\"\n",
    "    # import Bio Entrez, let NCBI know who I am\n",
    "\n",
    "    search_results = Entrez.read(Entrez.epost(db =\"protein\", \n",
    "                                              id = \",\".join(query_ids)))  # needs the id's to be as a list separated by commas\n",
    "    #upload the IDs to NCBI\n",
    "    webenv_return = search_results[\"WebEnv\"]\n",
    "    query_key_return = search_results[\"QueryKey\"]\n",
    "    #grab the relevant variables to call on the stuff we posted\n",
    "\n",
    "    count = len(query_ids)\n",
    "    #assigns the count based on the number of sequences we searched for\n",
    "\n",
    "    from urllib.error import HTTPError\n",
    "    # load required library for the try and except conditions\n",
    "\n",
    "    batch_size = 20\n",
    "    #this determines how many things we retrieve and write to the file\n",
    "    #can safely be larger in the future\n",
    "\n",
    "    out_filename = str(in_filename[:-7]+\".fasta\")\n",
    "    #attempting to rename the file based on the input of the original file\n",
    "    out_handle = open(out_filename, \"w\")\n",
    "    #open file to write to\n",
    "\n",
    "    for start in range(0, count, batch_size):\n",
    "        end = min(count, start+batch_size)\n",
    "        print(\"Going to download record %i to %i\" % (start+1, end))\n",
    "        attempt = 0\n",
    "        while attempt < 3:\n",
    "            attempt += 1\n",
    "            try:\n",
    "                fetch_handle = Entrez.efetch(db=\"protein\", #says which db\n",
    "                                             rettype=\"fasta\", #says what format the data should be in\n",
    "                                             retmode=\"text\", #what the output should be\n",
    "                                             retstart = start, #say what range of results you want returned\n",
    "                                             retmax = batch_size, #say end of range of results want returned\n",
    "                                             webenv = webenv_return, #specify the info we uploaded with ePost\n",
    "                                             query_key = query_key_return)\n",
    "            except HTTPError as err:\n",
    "                if 500 <= err.code <= 599:\n",
    "                    print(\"Recieved error from server %s\" % err)\n",
    "                    print(\"Attempt %i of 3\" % attempt)\n",
    "                    time.sleep\n",
    "                else:\n",
    "                    raise\n",
    "        data = fetch_handle.read()\n",
    "        fetch_handle.close()\n",
    "        out_handle.write(data)\n",
    "    out_handle.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align with MUSCLE in Biopython\n",
    "```\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "muscle_cline = MuscleCommandline(input = \"WHRN_tiny.fasta\", out = \"WHRN_tiny_Muscle_align.txt\")\n",
    "stdout, stder = muscle_cline()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run through RAxML\n",
    "\n",
    "in shell\n",
    "```\n",
    "./raxmlHPC -m PROTGAMMAWAG -p 12345 -s Path/to/alignment/WHRN_tiny_Muscle_align.txt  -# 5 -n WHRNtn2\n",
    "```\n",
    "path depends on where the alignment is stored"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Format the Sequence Names For Tree making\n",
    "\n",
    "\n",
    "Use Bash to grab the top lines of the Fasta lines\n",
    "\n",
    "```\n",
    "cat WHRN_tiny.fasta | grep \">\" > WHRN_tiny_annotations.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In python\n",
    "\n",
    "```\n",
    "WHRN_annotations = open(\"WHRN_tiny_annotations.txt\", \"r\").readlines()\n",
    "name_file= open(\"WHRN_gene&sci&annotation.txt\", \"w\")\n",
    "for line in WHRN_annotations:\n",
    "    #this handles formating of names\n",
    "    #there are essentially a number of formats the genes can be name in, os it makes the naming process a bit tricky\n",
    "    \n",
    "    sequence_id = line.split(\" \")[0].replace(\">\", \"\")\n",
    "    #split the sequence ID out and get rid of FASTA formating\n",
    "    \n",
    "    \n",
    "    formated_org = line.split(\"[\")[1].split(' ')\n",
    "    genus_code = formated_org[0][:3]\n",
    "    genus = formated_org[0]\n",
    "    species_code = formated_org[1][:3]\n",
    "    final_org_name = genus_code + \"_\" + species_code\n",
    "    #splits out the genus and species\n",
    "    \n",
    "    taxa_group = taxa_Dict[final_org_name]\n",
    "    #assigns the species to a broad taxonomic division \n",
    "    \n",
    "    gene_annotation_slice = line.split(\"[\")[0].split(\" \")[1:]\n",
    "    if \"PREDICTED:\" in gene_annotation_slice:\n",
    "        if \"-like\" in gene_annotation_slice:\n",
    "            gene_annotation = gene_annotation_slice[0].replace(\" \", \"\")\n",
    "        else:\n",
    "            gene_annotation = gene_annotation_slice[1:3]\n",
    "    elif \"Drosophila\" in gene_annotation_slice:\n",
    "        gene_annotation = gene_annotation_slice[1:3]\n",
    "    else:\n",
    "        gene_annotation = gene_annotation_slice[0:1]\n",
    "    gene_annotation_final = \"_\".join(gene_annotation).replace(\",\", \"\")\n",
    "    #cuts down the gene annoatation names to make them more manageable\n",
    "    #problematic as the gene names do not conform to a single format that makes them easily parsable\n",
    "    #have to switch between formats\n",
    "    \n",
    "    combined_name = '\"' +final_org_name + \" \" + gene_annotation_final + '\"'\n",
    "    genus_and_gene = '\"' +genus + \" \" + gene_annotation_final + '\"'\n",
    "    \n",
    "    name_file.write(sequence_id+\",\"+\n",
    "                    final_org_name+\",\"+\n",
    "                    gene_annotation_final+\",\"+ \n",
    "                    taxa_group+ \",\"+ \n",
    "                    combined_name+ \",\" +\n",
    "                    genus_and_gene+'\\n')\n",
    "    #Writes it to the file\n",
    "    \n",
    "name_file.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Tree in R\n",
    "\n",
    "```\n",
    "rm(list = ls())\n",
    "library(ape)\n",
    "### Basic Tree!\n",
    "\n",
    "WHRN_tree <- read.tree(\"/Path/o/best/Tree/RAxML_bestTree.WHRNtn2\")\n",
    "plot(WHRN_tree)\n",
    "#bring in the tree from where ever it is and plot it\n",
    "```\n",
    "<img src = \"files/WHRN_basic_tree.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "###### Increase the readability of the tree\n",
    "\n",
    "```\n",
    "# Read in and format the annotation data\n",
    "WHRN_gene_annotations <- read.csv(\"WHRN_gene&sci&annotation.txt\", \n",
    "                                  header =FALSE, \n",
    "                                  stringsAsFactors = FALSE)\n",
    "names(WHRN_gene_annotations) <- c(\"GeneID\", \"OrgName\", \"Annotation\", \"Group\", \"CombinedName\", \"Genus_Gene\")\n",
    "rownames(WHRN_gene_annotations) <- WHRN_gene_annotations$GeneID\n",
    "\n",
    "\n",
    "\n",
    "# Reorder data frame to match rows\n",
    "\n",
    "order <- match(WHRN_tree$tip.label, rownames(WHRN_gene_annotations))\n",
    "WHRN_gene_annotations <- WHRN_gene_annotations[,][order,]\n",
    "\n",
    "\n",
    "WHRN_tree$tip.label <- WHRN_gene_annotations$Genus_Gene\n",
    "# Rename tips of Tree to increase Readability\n",
    "\n",
    "plot(WHRN_tree, cex = 0.5, label.offset = 0.1)\n",
    "# decrease size of labels a bit\n",
    "\n",
    "title(\"Whirlin (USH2D) best hits tree\")\n",
    "#add a title\n",
    "add.scale.bar()\n",
    "#add a scale bar!\n",
    "\n",
    "# Color the tips to indicate which lineage the organism belongs to\n",
    "\n",
    "WHRN_gene_annotations$Group <- as.factor(WHRN_gene_annotations$Group)\n",
    "\n",
    "#start by setting the Group to factor\n",
    "lineages <- unique(WHRN_gene_annotations$Group)\n",
    "cols <- rainbow(n = length(lineages))\n",
    "#color vector for legend\n",
    "colvec <- cols[WHRN_gene_annotations$Group]\n",
    "#color vector for tree\n",
    "tiplabels(pch = 19, col = colvec)\n",
    "\n",
    "\n",
    "# Add a legend!\n",
    "legend(x = \"bottomright\", lwd =0, pch = 19 , legend = levels(lineages), col = cols, cex = 0.7)\n",
    "```\n",
    "Here is the output:\n",
    "\n",
    "\n",
    "\n",
    "<img src =\"files/Whirlin_presentation_color_tree.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Check for Homology\n",
    "use HMMER to annotate domains of sequences and see if the appropriate domains are there\n",
    "\n",
    "```\n",
    "for gene in list_of_retrieved_sequences:\n",
    "    hmmer gene vs pfam database\n",
    "        for hit from hmmer of sufficient significance:\n",
    "            add domain id and length to metadata of sequence\n",
    "        if annotated sequence has at least one of each domain present in search sequence:\n",
    "            add to document containing list of potential homologs and their sequence\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Reconcile tree with species tree\n",
    "compare the trees generated to established species trees\n",
    "```\n",
    "for gene_tree in list_of_tree_objects:\n",
    "    comparison_tree = (well support tree of Metazoa)\n",
    "    calibrate gene_tree with comparison_tree\n",
    "    \n",
    "```\n",
    "<img src=\"Darwins_first_tree.jpg\">\n",
    "\n",
    "Placeholder image of tree,  by Charles Darwin, retrieved from Wikimedia Commons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "Kremer, H., van Wijk, E., Märker, T., Wolfrum, U., & Roepman, R. (2006). Usher syndrome: molecular links of pathogenesis, proteins and pathways. Human molecular genetics, 15(suppl 2), R262-R270.\n",
    "\n",
    "Sodergren, E., Weinstock, G.M., Davidson, E.H., Cameron, R.A., Gibbs, R.A., Angerer, R.C., Angerer, L.M., Arnone, M.I., Burgess, D.R., Burke, R.D. and Coffman, J.A., 2006. The genome of the sea urchin Strongylocentrotus purpuratus. Science, 314(5801), pp.941-952.\n",
    "\n",
    "Peña, J. F., Alié, A., Richter, D. J., Wang, L., Funayama, N., & Nichols, S. A. (2016). Conserved expression of vertebrate microvillar gene homologs in choanocytes of freshwater sponges. EvoDevo, 7(1), 13."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
